race build is broken at 3fd320621566c44aad0f7df5d0f5858efd2bbcf6

It seems that the stdlib building is busted when `--features=race` is added to `bazel test`. 

Getting different errors on different platforms, but one is:

```
ERROR: /Users/jmhodges/founding/src/founding/go/cmd/https-sidecar/BUILD:4:1: GoCompile go/cmd/https-sidecar/~normal~go_default_library~/founding/go/cmd/https-sidecar.a failed (Exit 1).
go/cmd/https-sidecar/main.go:6:2: can't find import: "flag"
```

Same here (same macOS, same bazel, race enabled). Also prints a whole bunch of warnings:
```console
INFO: From SkylarkAction external/go_stdlib_darwin_amd64_cgo_race/bin/go [for host]:
# runtime/cgo
clang: warning: argument unused during compilation: '-pthread' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-gno-record-gcc-switches' [-Wunused-command-line-argument]
# runtime/cgo
clang: warning: argument unused during compilation: '-pthread' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-gno-record-gcc-switches' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-nopie' [-Wunused-command-line-argument]
# runtime/race
clang: warning: argument unused during compilation: '-pthread' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-gno-record-gcc-switches' [-Wunused-command-line-argument]
# runtime/race
clang: warning: argument unused during compilation: '-pthread' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-gno-record-gcc-switches' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-nopie' [-Wunused-command-line-argument]
# plugin
clang: warning: argument unused during compilation: '-pthread' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-gno-record-gcc-switches' [-Wunused-command-line-argument]
# plugin
clang: warning: argument unused during compilation: '-pthread' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-gno-record-gcc-switches' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-nopie' [-Wunused-command-line-argument]
# os/user
clang: warning: argument unused during compilation: '-pthread' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-gno-record-gcc-switches' [-Wunused-command-line-argument]
# os/user
clang: warning: argument unused during compilation: '-pthread' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-gno-record-gcc-switches' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-nopie' [-Wunused-command-line-argument]
# net
clang: warning: argument unused during compilation: '-pthread' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-gno-record-gcc-switches' [-Wunused-command-line-argument]
# net
clang: warning: argument unused during compilation: '-pthread' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-gno-record-gcc-switches' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-nopie' [-Wunused-command-line-argument]
# crypto/x509
clang: warning: argument unused during compilation: '-pthread' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-gno-record-gcc-switches' [-Wunused-command-line-argument]
# crypto/x509
clang: warning: argument unused during compilation: '-pthread' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-gno-record-gcc-switches' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-nopie' [-Wunused-command-line-argument]
ERROR: blabla/vendor/github.com/pmezard/go-difflib/difflib/BUILD.bazel:3:1: GoCompile vendor/github.com/pmezard/go-difflib/difflib/~normal~go_default_library~/github.com/pmezard/go-difflib/difflib.a failed (Exit 1).
vendor/github.com/pmezard/go-difflib/difflib/difflib.go:19:2: can't find import: "bufio"
2017/10/27 19:28:53 error running compiler: exit status 1
```
However it also fails on linux (travis) - https://travis-ci.org/atlassian/smith/builds/293558934
```
      Running (linux-sandbox):
        SkylarkAction external/go_stdlib_linux_amd64_cgo_race/bin/go [for host], 15 s
ERROR: (10-27 08:21:35.006) /home/travis/gopath/src/github.com/atlassian/smith/vendor/github.com/pmezard/go-difflib/difflib/BUILD.bazel:3:1: GoCompile vendor/github.com/pmezard/go-difflib/difflib/~normal~go_default_library~/github.com/pmezard/go-difflib/difflib.a failed (Exit 1): compile failed: error executing command 
  (cd /home/travis/.cache/bazel/_bazel_travis/21b3068201190d7334091ffc6389815b/execroot/__main__ && \
  exec env - \
  bazel-out/host/bin/external/io_bazel_rules_go/go/tools/builders/compile -go bazel-out/host/bin/external/go_stdlib_linux_amd64_cgo_race/bin/go -root_file bazel-out/host/bin/external/go_stdlib_linux_amd64_cgo_race/ROOT -goos linux -goarch amd64 '-cgo=1' -package_list external/go_sdk/packages.txt -src vendor/github.com/pmezard/go-difflib/difflib/difflib.go -o 'bazel-out/local-fastbuild/bin/vendor/github.com/pmezard/go-difflib/difflib/~normal~go_default_library~/github.com/pmezard/go-difflib/difflib.a' -trimpath . -I . -- -p github.com/pmezard/go-difflib/difflib)
Use --sandbox_debug to see verbose messages from the sandbox.
vendor/github.com/pmezard/go-difflib/difflib/difflib.go:19:2: can't find import: "bufio"
2017/10/27 08:21:34 error running compiler: exit status 1
```

--- a/go/private/actions/action.bzl
+++ b/go/private/actions/action.bzl
@@ -12,15 +12,16 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-def action_with_go_env(ctx, go_toolchain, stdlib, executable = None, command=None, arguments = [], inputs = [], **kwargs):
+def action_with_go_env(ctx, go_toolchain, mode, executable = None, command=None, arguments = [], inputs = [], **kwargs):
   if command:
     fail("You cannot run action_with_go_env with a 'command', only an 'executable'")
+  stdlib = go_toolchain.stdlib.get(ctx, go_toolchain, mode)
   args = [
       "-go", stdlib.go.path,
       "-root_file", stdlib.root_file.path,
       "-goos", stdlib.goos,
       "-goarch", stdlib.goarch,
-      "-cgo=" + ("1" if stdlib.cgo else "0"),
+      "-cgo=" + ("0" if mode.pure else "1"),
   ] + arguments
   ctx.action(
       inputs = depset(inputs) + stdlib.files,

--- a/go/private/actions/asm.bzl
+++ b/go/private/actions/asm.bzl
@@ -12,6 +12,9 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+load("@io_bazel_rules_go//go/private:mode.bzl",
+    "NORMAL_MODE",
+)
 load("@io_bazel_rules_go//go/private:actions/action.bzl",
     "action_with_go_env",
 )
@@ -19,19 +22,20 @@ load("@io_bazel_rules_go//go/private:actions/action.bzl",
 def emit_asm(ctx, go_toolchain,
     source = None,
     hdrs = [],
-    out_obj = None):
+    out_obj = None,
+    mode = NORMAL_MODE):
   """See go/toolchains.rst#asm for full documentation."""
 
   if source == None: fail("source is a required parameter")
   if out_obj == None: fail("out_obj is a required parameter")
-  stdlib = go_toolchain.stdlib.get(ctx, go_toolchain)
+  stdlib = go_toolchain.stdlib.get(ctx, go_toolchain, mode)
   includes = depset([stdlib.root_file.dirname + "/pkg/include"])
   includes += [f.dirname for f in hdrs]
   inputs = hdrs + stdlib.files + [source]
   asm_args = [source.path, "-o", out_obj.path]
   for inc in includes:
     asm_args += ["-I", inc]
-  action_with_go_env(ctx, go_toolchain, stdlib,
+  action_with_go_env(ctx, go_toolchain, mode,
       inputs = list(inputs),
       outputs = [out_obj],
       mnemonic = "GoAsmCompile",

--- a/go/private/actions/binary.bzl
+++ b/go/private/actions/binary.bzl
@@ -12,8 +12,9 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-load("@io_bazel_rules_go//go/private:common.bzl",
-    "link_modes",
+load("@io_bazel_rules_go//go/private:mode.bzl",
+    "mode_string",
+    "common_modes",
     "NORMAL_MODE",
     "RACE_MODE",
     "STATIC_MODE",
@@ -51,29 +52,34 @@ def emit_binary(ctx, go_toolchain,
   executables = {}
   extension = "" # TODO: .exe on windows
 
-  if default:
-    executables["default"] = default
-
-  for mode in link_modes:
-    executable = ctx.new_file(name + "." + mode + extension)
-    executables[mode] = executable
-
-  for mode, executable in executables.items():
-    if mode == "default":
-        # work out what the default mode should be
-        if "race" in ctx.features:
-            mode = RACE_MODE
-        elif "static" in ctx.features:
-            mode = STATIC_MODE
-        else:
-            mode = NORMAL_MODE
-
+  for mode in common_modes:
+    executable = ctx.new_file(name + "." + mode_string(mode) + extension)
+    executables[mode_string(mode)] = executable
     go_toolchain.actions.link(
         ctx,
         go_toolchain = go_toolchain,
         library=golib,
         mode=mode,
         executable=executable,
+        gc_linkopts=gc_linkopts,
+        x_defs=x_defs,
+    )
+
+  if default:
+    executables["default"] = default
+    # work out what the default mode should be
+    if "race" in ctx.features:
+        mode = RACE_MODE
+    elif "static" in ctx.features:
+        mode = STATIC_MODE
+    else:
+        mode = NORMAL_MODE
+    go_toolchain.actions.link(
+        ctx,
+        go_toolchain = go_toolchain,
+        library=golib,
+        mode=mode,
+        executable=default,
         gc_linkopts=gc_linkopts,
         x_defs=x_defs,
     )

--- a/go/private/actions/binary.bzl
+++ b/go/private/actions/binary.bzl
@@ -15,6 +15,7 @@
 load("@io_bazel_rules_go//go/private:mode.bzl",
     "mode_string",
     "common_modes",
+    "get_mode",
     "NORMAL_MODE",
     "RACE_MODE",
     "STATIC_MODE",
@@ -67,13 +68,7 @@ def emit_binary(ctx, go_toolchain,
 
   if default:
     executables["default"] = default
-    # work out what the default mode should be
-    if "race" in ctx.features:
-        mode = RACE_MODE
-    elif "static" in ctx.features:
-        mode = STATIC_MODE
-    else:
-        mode = NORMAL_MODE
+    mode = get_mode(ctx)
     go_toolchain.actions.link(
         ctx,
         go_toolchain = go_toolchain,

--- a/go/private/actions/compile.bzl
+++ b/go/private/actions/compile.bzl
@@ -12,9 +12,8 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-load("@io_bazel_rules_go//go/private:common.bzl",
+load("@io_bazel_rules_go//go/private:mode.bzl",
     "NORMAL_MODE",
-    "RACE_MODE",
 )
 load("@io_bazel_rules_go//go/private:providers.bzl",
     "get_library",
@@ -37,11 +36,11 @@ def emit_compile(ctx, go_toolchain,
   if sources == None: fail("sources is a required parameter")
   if out_lib == None: fail("out_lib is a required parameter")
 
-  stdlib = go_toolchain.stdlib.get(ctx, go_toolchain)
-
   # Add in any mode specific behaviours
-  if mode == RACE_MODE:
+  if mode.race:
     gc_goopts = gc_goopts + ("-race",)
+  if mode.msan:
+    gc_goopts = gc_goopts + ("-msan",)
 
   gc_goopts = [ctx.expand_make_variables("gc_goopts", f, {}) for f in gc_goopts]
   inputs = sources + [go_toolchain.data.package_list]
@@ -63,7 +62,7 @@ def emit_compile(ctx, go_toolchain,
   if ctx.attr._go_toolchain_flags.compilation_mode == "debug":
     args.extend(["-N", "-l"])
   args.extend(cgo_sources)
-  action_with_go_env(ctx, go_toolchain, stdlib,
+  action_with_go_env(ctx, go_toolchain, mode,
       inputs = list(inputs),
       outputs = [out_lib],
       mnemonic = "GoCompile",

--- a/go/private/actions/cover.bzl
+++ b/go/private/actions/cover.bzl
@@ -12,15 +12,18 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+load("@io_bazel_rules_go//go/private:mode.bzl",
+    "NORMAL_MODE",
+)
 load("@io_bazel_rules_go//go/private:actions/action.bzl",
     "action_with_go_env",
 )
 
 def emit_cover(ctx, go_toolchain,
-               sources = []):
+               sources = [],
+               mode = NORMAL_MODE):
   """See go/toolchains.rst#cover for full documentation."""
 
-  stdlib = go_toolchain.stdlib.get(ctx, go_toolchain)
   outputs = []
   # TODO(linuxerwang): make the mode configurable.
   cover_vars = []
@@ -36,7 +39,7 @@ def emit_cover(ctx, go_toolchain,
     cover_vars += ["{}={}".format(cover_var,src.short_path)]
     out = ctx.new_file(cover_var + '.cover.go')
     outputs += [out]
-    action_with_go_env(ctx, go_toolchain, stdlib,
+    action_with_go_env(ctx, go_toolchain, mode,
         inputs = [src],
         outputs = [out],
         mnemonic = "GoCover",

--- a/go/private/actions/library.bzl
+++ b/go/private/actions/library.bzl
@@ -16,7 +16,10 @@ load("@io_bazel_rules_go//go/private:common.bzl",
     "dict_of",
     "split_srcs",
     "join_srcs",
-    "compile_modes",
+)
+load("@io_bazel_rules_go//go/private:mode.bzl",
+    "common_modes",
+    "mode_string",
 )
 load("@io_bazel_rules_go//go/private:providers.bzl",
     "CgoInfo",
@@ -94,8 +97,8 @@ def emit_library(ctx, go_toolchain,
   lib_name = importpath + ".a"
   compilepath = importpath if importable else None
   mode_fields = {} # These are added to the GoLibrary provider directly
-  for mode in compile_modes:
-    out_dir = "~{}~{}~".format(mode, ctx.label.name)
+  for mode in common_modes:
+    out_dir = "~{}~{}~".format(mode_string(mode), ctx.label.name)
     out_lib = ctx.new_file("{}/{}".format(out_dir, lib_name))
     searchpath = out_lib.path[:-len(lib_name)]
     mode_fields[library_attr(mode)] = out_lib

--- a/go/private/actions/link.bzl
+++ b/go/private/actions/link.bzl
@@ -12,10 +12,9 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-load("@io_bazel_rules_go//go/private:common.bzl",
+load("@io_bazel_rules_go//go/private:mode.bzl",
     "NORMAL_MODE",
-    "RACE_MODE",
-    "STATIC_MODE",
+    "LINKMODE_NORMAL",
 )
 load("@io_bazel_rules_go//go/private:providers.bzl",
     "get_library",
@@ -37,7 +36,7 @@ def emit_link(ctx, go_toolchain,
   if library == None: fail("library is a required parameter")
   if executable == None: fail("executable is a required parameter")
 
-  stdlib = go_toolchain.stdlib.get(ctx, go_toolchain)
+  stdlib = go_toolchain.stdlib.get(ctx, go_toolchain, mode)
 
   config_strip = len(ctx.configuration.bin_dir.path) + 1
   pkg_depth = executable.dirname[config_strip:].count('/') + 1
@@ -51,21 +50,23 @@ def emit_link(ctx, go_toolchain,
 
   gc_linkopts, extldflags = _extract_extldflags(gc_linkopts, extldflags)
 
-  libmode = mode
   # Add in any mode specific behaviours
-  if mode == RACE_MODE:
+  if mode.race:
     gc_linkopts += ["-race"]
-  elif mode == STATIC_MODE:
-    libmode = NORMAL_MODE
+  if mode.msan:
+    gc_linkopts += ["-msan"]
+  if mode.static:
     gc_linkopts = gc_linkopts + ["-linkmode", "external"]
     extldflags.append("-static")
+  if mode.link != LINKMODE_NORMAL:
+    fail("Link mode {} is not yet supported".format(mode.link))
 
   link_opts = ["-L", "."]
   libs = depset()
   cgo_deps = depset()
   for golib in depset([library]) + library.transitive:
-    libs += [get_library(golib, libmode)]
-    link_opts += ["-L", get_searchpath(golib, libmode)]
+    libs += [get_library(golib, mode)]
+    link_opts += ["-L", get_searchpath(golib, mode)]
     cgo_deps += golib.cgo_deps
 
   for d in cgo_deps:
@@ -96,7 +97,7 @@ def emit_link(ctx, go_toolchain,
         "-extld", ld,
         "-extldflags", " ".join(extldflags),
     ]
-  link_opts += [get_library(golib, libmode).path]
+  link_opts += [get_library(golib, mode).path]
   link_args = []
   # Stamping support
   stamp_inputs = []
@@ -114,7 +115,7 @@ def emit_link(ctx, go_toolchain,
 
   link_args += ["--"] + link_opts
 
-  action_with_go_env(ctx, go_toolchain, stdlib,
+  action_with_go_env(ctx, go_toolchain, mode,
       inputs = list(libs + cgo_deps +
                 go_toolchain.data.crosstool + stamp_inputs),
       outputs = [executable],

--- a/go/private/actions/pack.bzl
+++ b/go/private/actions/pack.bzl
@@ -12,6 +12,9 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+load("@io_bazel_rules_go//go/private:mode.bzl",
+    "NORMAL_MODE",
+)
 load("@io_bazel_rules_go//go/private:actions/action.bzl",
     "action_with_go_env",
 )
@@ -20,13 +23,12 @@ def emit_pack(ctx, go_toolchain,
     in_lib = None,
     out_lib = None,
     objects = (),
-    archive = None):
+    archive = None,
+    mode = NORMAL_MODE):
   """See go/toolchains.rst#pack for full documentation."""
 
   if in_lib == None: fail("in_lib is a required parameter")
   if out_lib == None: fail("out_lib is a required parameter")
-
-  stdlib = go_toolchain.stdlib.get(ctx, go_toolchain)
 
   inputs = [in_lib]
 
@@ -42,7 +44,7 @@ def emit_pack(ctx, go_toolchain,
     inputs.append(archive)
     arguments.extend(["-arc", archive.path])
 
-  action_with_go_env(ctx, go_toolchain, stdlib,
+  action_with_go_env(ctx, go_toolchain, mode,
       inputs = inputs,
       outputs = [out_lib],
       mnemonic = "GoPack",

--- a/go/private/common.bzl
+++ b/go/private/common.bzl
@@ -17,14 +17,6 @@ load("@io_bazel_rules_go//go/private:providers.bzl", "GoLibrary")
 DEFAULT_LIB = "go_default_library"
 VENDOR_PREFIX = "/vendor/"
 
-# Modes are documented in go/modes.rst#compilation-modes
-NORMAL_MODE = "normal"
-RACE_MODE = "race"
-STATIC_MODE = "static"
-
-compile_modes = (NORMAL_MODE, RACE_MODE)
-link_modes =  (NORMAL_MODE, RACE_MODE, STATIC_MODE)
-
 go_exts = [
     ".go",
 ]

--- a/go/private/go_toolchain.bzl
+++ b/go/private/go_toolchain.bzl
@@ -23,12 +23,12 @@ load("@io_bazel_rules_go//go/private:actions/link.bzl", "emit_link", "bootstrap_
 load("@io_bazel_rules_go//go/private:actions/pack.bzl", "emit_pack")
 load("@io_bazel_rules_go//go/private:providers.bzl", "GoStdLib")
 
-def _get_stdlib(ctx, go_toolchain):
-  if "pure" in ctx.features and "race" in ctx.features:
+def _get_stdlib(ctx, go_toolchain, mode):
+  if mode.race and mode.pure:
     return go_toolchain.stdlib.pure_race
-  elif "pure" in ctx.features:
+  elif mode.pure:
     return go_toolchain.stdlib.pure
-  elif "race" in ctx.features:
+  elif mode.race:
     return go_toolchain.stdlib.cgo_race
   else:
     return go_toolchain.stdlib.cgo

--- a/go/private/mode.bzl
+++ b/go/private/mode.bzl
@@ -1,0 +1,74 @@
+# Copyright 2014 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Modes are documented in go/modes.rst#compilation-modes
+
+LINKMODE_NORMAL = "normal"
+LINKMODE_SHARED = "shared"
+LINKMODE_PIE = "pie"
+LINKMODE_PLUGIN = "plugin"
+
+def mode(base=None, static=None, race=None, msan=None, pure=None, link=None):
+  if static == None: static = base.static
+  if race == None: race = base.race
+  if msan == None: msan = base.msan
+  if pure == None: pure = base.pure
+  if link == None: link = base.link
+  return struct(
+      static = static,
+      race = race,
+      msan = msan,
+      pure = pure,
+      link = link,
+  )
+
+DEFAULT_MODE = mode(
+    static = False,
+    race = False,
+    msan = False,
+    pure = False,
+    link = LINKMODE_NORMAL,
+)
+
+def mode_string(mode):
+  result = []
+  if mode.static:
+    result.append("static")
+  if mode.race:
+    result.append("race")
+  if mode.msan:
+    result.append("msan")
+  if mode.pure:
+    result.append("pure")
+  if not result or not mode.link == LINKMODE_NORMAL:
+    result.append(mode.link)
+  return "_".join(result)
+
+NORMAL_MODE = DEFAULT_MODE
+RACE_MODE = mode(base = DEFAULT_MODE, race=True)
+STATIC_MODE = mode(base = DEFAULT_MODE, static=True)
+
+common_modes = (NORMAL_MODE, RACE_MODE, STATIC_MODE)
+
+def get_mode(ctx):
+  #TODO: allow ctx.attr to override the feature driven defaults
+  #TODO: allow link mode selection
+  return mode(
+      base=DEFAULT_MODE,
+      static = True if "static" in ctx.features else False,
+      race = True if "race" in ctx.features else False,
+      msan = True if "msan" in ctx.features else False,
+      pure = True if "pure" not in ctx.features else False,
+      link = LINKMODE_NORMAL,
+  )

--- a/go/private/mode.bzl
+++ b/go/private/mode.bzl
@@ -1,4 +1,4 @@
-# Copyright 2014 The Bazel Authors. All rights reserved.
+# Copyright 2017 The Bazel Authors. All rights reserved.
 #
 # Licensed under the Apache License, Version 2.0 (the "License");
 # you may not use this file except in compliance with the License.
@@ -69,6 +69,6 @@ def get_mode(ctx):
       static = True if "static" in ctx.features else False,
       race = True if "race" in ctx.features else False,
       msan = True if "msan" in ctx.features else False,
-      pure = True if "pure" not in ctx.features else False,
+      pure = True if "pure" in ctx.features else False,
       link = LINKMODE_NORMAL,
   )

--- a/go/private/providers.bzl
+++ b/go/private/providers.bzl
@@ -12,6 +12,8 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+load("@io_bazel_rules_go//go/private:mode.bzl", "mode_string")
+
 GoLibrary = provider()
 """See go/providers.rst#GoLibrary for full documentation."""
 
@@ -29,30 +31,30 @@ GoStdLib = provider()
 def library_attr(mode):
   """Returns the attribute name for the library of the given mode.
 
-  mode must be one of the common.bzl#compile_modes
+  mode must a struct returned by common.bzl#mode
   """
-  return mode+"_library"
+  return mode_string(mode)+"_library"
 
 def get_library(golib, mode):
   """Returns the compiled library for the given mode
 
   golib must be a GoLibrary
-  mode must be one of the common.bzl#compile_modes
+  mode must a struct returned by common.bzl#mode
   """
   return getattr(golib, library_attr(mode))
 
 def searchpath_attr(mode):
   """Returns the search path for the given mode
 
-  mode must be one of the common.bzl#compile_modes
+  mode must a struct returned by common.bzl#mode
   """
-  return mode+"_searchpath"
+  return mode_string(mode)+"_searchpath"
 
 def get_searchpath(golib, mode):
   """Returns the search path for the given mode
 
   golib must be a GoLibrary
-  mode must be one of the common.bzl#compile_modes
+  mode must a struct returned by common.bzl#mode
   """
   return getattr(golib, searchpath_attr(mode))
 

--- a/go/private/rules/binary.bzl
+++ b/go/private/rules/binary.bzl
@@ -13,11 +13,8 @@
 # limitations under the License.
 
 load("@io_bazel_rules_go//go/private:common.bzl",
-    "compile_modes",
     "go_filetype",
     "go_importpath",
-    "NORMAL_MODE",
-    "RACE_MODE",
 )
 load("@io_bazel_rules_go//go/private:rules/prefix.bzl",
     "go_prefix_default",

--- a/go/private/rules/cgo.bzl
+++ b/go/private/rules/cgo.bzl
@@ -12,8 +12,19 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-load("@io_bazel_rules_go//go/private:common.bzl", "dict_of", "split_srcs", "join_srcs", "pkg_dir")
-load("@io_bazel_rules_go//go/private:providers.bzl", "CgoInfo", "GoLibrary")
+load("@io_bazel_rules_go//go/private:common.bzl",
+    "dict_of",
+    "split_srcs",
+    "join_srcs",
+    "pkg_dir"
+)
+load("@io_bazel_rules_go//go/private:mode.bzl",
+    "get_mode",
+)
+load("@io_bazel_rules_go//go/private:providers.bzl",
+    "CgoInfo",
+    "GoLibrary",
+)
 load("@io_bazel_rules_go//go/private:actions/action.bzl",
     "action_with_go_env",
 )
@@ -46,7 +57,8 @@ def _select_archive(files):
 
 def _cgo_codegen_impl(ctx):
   go_toolchain = ctx.toolchains["@io_bazel_rules_go//go:toolchain"]
-  stdlib = go_toolchain.stdlib.get(ctx, go_toolchain)
+  mode = get_mode(ctx)
+  stdlib = go_toolchain.stdlib.get(ctx, go_toolchain, mode)
   if not stdlib.cgo_tools:
     fail("Go toolchain does not support cgo")
   linkopts = ctx.attr.linkopts[:]
@@ -108,7 +120,7 @@ def _cgo_codegen_impl(ctx):
   # The first -- below is to stop the cgo from processing args, the
   # second is an actual arg to forward to the underlying go tool
   args += ["--", "--"] + copts
-  action_with_go_env(ctx, go_toolchain, stdlib,
+  action_with_go_env(ctx, go_toolchain, mode,
       inputs = inputs,
       outputs = list(c_outs + go_outs + [cgo_main]),
       mnemonic = "CGoCodeGen",
@@ -155,14 +167,14 @@ _cgo_codegen = rule(
 
 def _cgo_import_impl(ctx):
   go_toolchain = ctx.toolchains["@io_bazel_rules_go//go:toolchain"]
-  stdlib = go_toolchain.stdlib.get(ctx, go_toolchain)
+  mode = get_mode(ctx)
   args = [
       "-dynout", ctx.outputs.out.path,
       "-dynimport", ctx.file.cgo_o.path,
       "-src", ctx.files.sample_go_srcs[0].path,
   ]
 
-  action_with_go_env(ctx, go_toolchain, stdlib,
+  action_with_go_env(ctx, go_toolchain, mode,
       inputs = [
           ctx.file.cgo_o,
           ctx.files.sample_go_srcs[0],

--- a/go/private/rules/info.bzl
+++ b/go/private/rules/info.bzl
@@ -12,15 +12,18 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+load("@io_bazel_rules_go//go/private:mode.bzl",
+    "get_mode",
+)
 load("@io_bazel_rules_go//go/private:actions/action.bzl",
     "action_with_go_env",
 )
 
 def _go_info_script_impl(ctx):
   go_toolchain = ctx.toolchains["@io_bazel_rules_go//go:toolchain"]
-  stdlib = go_toolchain.stdlib.get(ctx, go_toolchain)
+  mode = get_mode(ctx)
   out = ctx.actions.declare_file(ctx.label.name+".bash")
-  action_with_go_env(ctx, go_toolchain, stdlib,
+  action_with_go_env(ctx, go_toolchain, mode,
       inputs = [],
       outputs = [out],
       mnemonic = "GoInfo",

--- a/go/private/rules/library.bzl
+++ b/go/private/rules/library.bzl
@@ -15,6 +15,8 @@
 load("@io_bazel_rules_go//go/private:common.bzl",
     "go_filetype",
     "go_importpath",
+)
+load("@io_bazel_rules_go//go/private:mode.bzl",
     "RACE_MODE",
     "NORMAL_MODE",
 )

--- a/go/private/rules/stdlib.bzl
+++ b/go/private/rules/stdlib.bzl
@@ -109,7 +109,6 @@ def _stdlib_impl(ctx):
           goarch = ctx.attr.goarch,
           libs = [pkg],
           headers = [pkg],
-          cgo = ctx.attr.cgo,
           files = files,
           cgo_tools = struct(
               compiler_executable = cpp.compiler_executable,

--- a/go/private/rules/test.bzl
+++ b/go/private/rules/test.bzl
@@ -17,8 +17,9 @@ load("@io_bazel_rules_go//go/private:common.bzl",
     "go_importpath",
     "split_srcs",
     "pkg_dir",
-    "NORMAL_MODE",
-    "RACE_MODE",
+)
+load("@io_bazel_rules_go//go/private:mode.bzl",
+    "get_mode",
 )
 load("@io_bazel_rules_go//go/private:rules/prefix.bzl",
     "go_prefix_default",
@@ -41,7 +42,7 @@ def _go_test_impl(ctx):
   test into a binary."""
 
   go_toolchain = ctx.toolchains["@io_bazel_rules_go//go:toolchain"]
-  stdlib = go_toolchain.stdlib.get(ctx, go_toolchain)
+  mode = get_mode(ctx)
   embed = ctx.attr.embed
   if ctx.attr.library:
     embed = embed + [ctx.attr.library]
@@ -85,7 +86,7 @@ def _go_test_impl(ctx):
       for var in g.cover_vars:
         arguments += ["-cover", "{}={}".format(var, g.importpath)]
 
-  action_with_go_env(ctx, go_toolchain, stdlib,
+  action_with_go_env(ctx, go_toolchain, mode,
       inputs = go_srcs,
       outputs = [main_go],
       mnemonic = "GoTestGenTest",

--- a/go/private/tools/vet.bzl
+++ b/go/private/tools/vet.bzl
@@ -14,13 +14,18 @@
 
 load("@io_bazel_rules_go//go/private:providers.bzl", "GoPath")
 
+load("@io_bazel_rules_go//go/private:mode.bzl",
+    "get_mode",
+)
+
 def _go_vet_generate_impl(ctx):
   print("""
 EXPERIMENTAL: the go_vet_test rule is still very experimental
 Please do not rely on it for production use, but feel free to use it and file issues
 """)
   go_toolchain = ctx.toolchains["@io_bazel_rules_go//go:toolchain"]
-  stdlib = go_toolchain.stdlib.get(ctx, go_toolchain)
+  mode = get_mode(ctx)
+  stdlib = go_toolchain.stdlib.get(ctx, go_toolchain, mode)
   script_file = ctx.new_file(ctx.label.name+".bash")
   gopath = []
   files = ctx.files.data + stdlib.files

--- a/go/private/tools/vet.bzl
+++ b/go/private/tools/vet.bzl
@@ -15,7 +15,7 @@
 load("@io_bazel_rules_go//go/private:providers.bzl", "GoPath")
 
 load("@io_bazel_rules_go//go/private:mode.bzl",
-    "get_mode",
+    "NORMAL_MODE",
 )
 
 def _go_vet_generate_impl(ctx):
@@ -24,7 +24,7 @@ EXPERIMENTAL: the go_vet_test rule is still very experimental
 Please do not rely on it for production use, but feel free to use it and file issues
 """)
   go_toolchain = ctx.toolchains["@io_bazel_rules_go//go:toolchain"]
-  mode = get_mode(ctx)
+  mode = NORMAL_MODE
   stdlib = go_toolchain.stdlib.get(ctx, go_toolchain, mode)
   script_file = ctx.new_file(ctx.label.name+".bash")
   gopath = []

--- a/proto/def.bzl
+++ b/proto/def.bzl
@@ -1,5 +1,7 @@
 load("@io_bazel_rules_go//go/private:common.bzl",
     "go_importpath",
+)
+load("@io_bazel_rules_go//go/private:mode.bzl",
     "RACE_MODE",
     "NORMAL_MODE",
 )

